executor/hash_table.go

```diff
@@ -14,31 +14,175 @@
 package executor
 
 import (
+	"hash"
+	"hash/fnv"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/memory"
 )
 
-const maxEntrySliceLen = 8 * 1024
+// hashRowContainer handles the rows and the hash map of a table.
+// TODO: support spilling out to disk when memory is limited.
+type hashRowContainer struct {
+	records   *chunk.List
+	hashTable *rowHashMap
+
+	sc        *stmtctx.StatementContext
+	allTypes  []*types.FieldType
+	keyColIdx []int
+	h         hash.Hash64
+	buf       [1]byte
+}
+
+func newHashRowContainer(
+	sc *stmtctx.StatementContext, statCount int,
+	allTypes []*types.FieldType, keyColIdx []int, initCap, maxChunkSize int) *hashRowContainer {
+
+	c := &hashRowContainer{
+		hashTable: newRowHashMapWithStatCount(statCount),
+		sc:        sc,
+		allTypes:  allTypes,
+		keyColIdx: keyColIdx,
+		h:         fnv.New64(),
+	}
+	c.records = chunk.NewList(allTypes, initCap, maxChunkSize)
+	return c
+}
+
+func (c *hashRowContainer) GetMemTracker() *memory.Tracker {
+	return c.records.GetMemTracker()
+}
+
+// GetMatchedRows get matched rows from probeRow. It can be called
+// in multiple goroutines while each goroutine should keep its own
+// h and buf.
+func (c *hashRowContainer) GetMatchedRows(probeRow chunk.Row, joinKeysTypes []*types.FieldType, keyColIdx []int, h hash.Hash64, buf []byte) (matched []chunk.Row, hasNull bool, err error) {
+
+	var key uint64
+	hasNull, key, err = c.getJoinKeyFromChkRow(c.sc, probeRow, joinKeysTypes, keyColIdx, h, buf)
+	if err != nil {
+		return
+	}
+	if hasNull {
+		return
+	}
+	innerPtrs := c.hashTable.Get(key)
+	if len(innerPtrs) == 0 {
+		hasNull = true
```

Comment:
Why we set `hashNull` to `true`?

Reply:
It means no inner rows matched by the outer row in `hashRowContainer` so we need call `onMissMatch`.
The same logic in old code is at https://github.com/pingcap/tidb/blob/master/executor/join.go#L417 and https://github.com/pingcap/tidb/blob/master/executor/join.go#L434

Reply:
I've addressed this by removing the ambiguous `isNull` value, and use the size of matched rows to call `onMissMatch`.